File: src/tokenBucket.ts
class TokenBucket {
    private tokens: number;
    private lastRefill: number;//last refill oda time

    constructor(
        private capacity: number,
        private refillRate: number
    ) {
        this.tokens = capacity;
        this.lastRefill = Date.now();
    }

    allowRequest(): boolean {
        const now = Date.now();

        const elapsedSeconds = (now - this.lastRefill) / 1000;

        const newTokens = elapsedSeconds * this.refillRate;

        this.tokens = Math.min(
            this.capacity,
            this.tokens + newTokens
        );

        this.lastRefill = now;

        if (this.tokens >= 1) {
            this.tokens -= 1;
            return true;
        }

        return false;
    }
}

export default TokenBucket;

File: src/tokenBucket.lua
local key = KEYS[1] 

local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

local data = redis.call("HMGET", key, "tokens", "last_refill")

local tokens = tonumber(data[1])
local last_refill = tonumber(data[2])

if tokens == nil then
    tokens = capacity
    last_refill = now
end

local elapsed = (now - last_refill) / 1000

tokens = math.min(
    capacity,
    tokens + (elapsed * refill_rate)
)

local allowed = 0

if tokens >= 1 then
    tokens = tokens - 1
    allowed = 1
end

redis.call(
    "HSET",
    key,
    "tokens", tokens,
    "last_refill", now
)

return { allowed, tokens }